Model/src/main/java/org/gusdb/wdk/model/user/dataset/event/UserDatasetEventCleanup.java

```diff
@@ -0,0 +1,84 @@
+package org.gusdb.wdk.model.user.dataset.event;
+
+import javax.sql.DataSource;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.gusdb.wdk.model.WdkModelException;
+import org.gusdb.wdk.model.user.dataset.UnsupportedTypeHandler;
+import org.gusdb.wdk.model.user.dataset.event.model.UserDatasetUninstallEvent;
+
+public class UserDatasetEventCleanup extends UserDatasetEventProcessor
+{
+  private static final Logger LOG = LogManager.getLogger(UserDatasetEventCleanup.class);
+
+  public UserDatasetEventCleanup(String projectID) throws WdkModelException {
+    super(projectID);
+  }
+
+  public void cleanupFailedInstalls() throws WdkModelException {
+    LOG.info("Beginning user dataset event cleanup.");
+
+    try (var appDb = openAppDB()) {
+
+      var handler = initHandler(appDb.getDataSource());
+      var events  = handler.getCleanableEvents();
+
+      for (var event : events) {
+        LOG.info("Processing event: " + event.getEventID());
+
+        if (!handler.shouldHandleEvent(event)) {
+          LOG.info("Skipping event");
+          continue;
+        }
+
+        if (!handler.acquireEventLock(event)) {
+          LOG.info("Event already claimed.");
+          continue;
+        }
+
+        // Error recovery block.  Errors in this try/catch do not halt event
+        // processing.  Instead when an error occurs, the event will be marked
+        // with an error status and the process will continue.
+        try {
+          var typeHandler = getUserDatasetStore().getTypeHandler(event.getUserDatasetType());
+
+          // A type handler was removed after the event was "installed".  This
+          // should never happen, but safety first.
+          if (UnsupportedTypeHandler.NAME.equals(typeHandler.getUserDatasetType().getName())) {
+            var error = "Type handler for type " + event.getUserDatasetType().getName() + " has been removed.";
+            LOG.error(error + "  Marking cleanup as failed.");
+            handler.failEvent(event, new Exception(error));
+            continue;
+          }
+
+          LOG.info("Attempting to run event cleanup.");
+
+          handler.handleUninstallEvent(new UserDatasetUninstallEvent(
```

Comment:
I'm noticing this is also run on e.g. a cleanup-ready share event. I have no idea what it means to clean up a share event - I guess they don't exist?

I had a concern about concurrency but then I realised `shouldHandleEvent` will be false if another event for the same UD has started processing.

Reply:
I have no idea what's involved in installing/uninstalling any event.  If share events have no uninstall process i guess it could be skipped here, or have a no-op in the underlying type handler?

Reply:
@Foxcapades only installs and uninstalls cause clean-up-able situations because they run the custom plugins that depend on a dataset type. But, like, a question of whether to uninstall when cleaning up a share process only comes up because the clean-ups are modelled as per event in the database.
Similarly it's a bit strange that the code will run an uninstall multiple times when there are multiple events marked for clean-up.

To me this means the status field would really belong in ApiDBUserDatasets.InstalledUserDataset but it's more like, a comment for https://github.com/VEuPathDB/ApiCommonData/pull/7 , and it would involve changing this PR by so much that it feels pointless to even say it unless it sounds like an inspiring idea to @steve-fischer-200 or we can think of some bugs this causes. 

Which I might do! But unnecessary or multiple uninstalls are not a bug because the uninstaller should be able to handle that.

Reply:
Actually I've realised last night cleaning up a share event might cause a following bug.

Say there are "install A", "share A". I set "share A" to clean up ready, run the cleaner, then run the sync. I think this will leave A uninstalled, so the sync does not bring the database into a correct state.

Reply:
To reply to your messages:

> Similarly it's a bit strange that the code will run an uninstall multiple times when there are multiple events marked for clean-up.

I'm not sure I'm understanding this correctly, but I will respond to what I think you're getting at.

Since the process will stop processing all events for a UD after a single failure, there should never be more than one "cleanup_ready" event for any given UD.

> To me this means the status field would really belong in ApiDBUserDatasets.InstalledUserDataset

My concern with this is how would we record event failures for an event that errors before it reaches the InstalledUserDataset table?  Additionally how would this work for failed share or uninstall events?

> Say there are "install A", "share A". I set "share A" to clean up ready, run the cleaner, then run the sync. I think this will leave A uninstalled, so the sync does not bring the database into a correct state.

After investigating the call chain a little bit, I agree, "share" events should not be "uninstalled", but should they have any processing done?  Such as revoking shares or something?

Reply:
Thank you for considering this!

> Since the process will stop processing all events for a UD after a single failure, there should never be more than one "cleanup_ready" event for any given UD.
I could set multiple successfully processed events for the same UD to cleanup_ready, and get the multiple uninstalls back.

> how would we record event failures for an event that errors before it reaches the InstalledUserDataset table?

UserDatasetEventSyncHandler.java does:
```
    // insert into the installedTable
    getInstallRepo().insertUserDataset(datasetId, userDataset.getMeta().getName());

    // insert into the type-specific tables
    typeHandler.installInAppDb(userDataset, cwd, getProjectId(), files);
    typeHandler.deleteWorkingDir(cwd);

    // grant access to the owner, by installing into the ownerTable
    grantAccess(ownerUserId, datasetId);
```
We would first insert the dataset with a processing status, and change it to successful or failed based on the results of `installInAppDb`. For uninstalls, similarly, first change the status to processing (or a new uninstall_processing), and afterwards either delete the entry on successful uninstall or change the status to failed (or a new uninstall_failed).

> how would this work for failed share or uninstall events?
In my mind, actions that may need cleaning up correspond to calls to `installInAppDb` and `uninstallInAppDb` - they might go wrong, and leave behind bad data. Shares don't go wrong, and cleaning after an install and uninstall is the same - by running an uninstaller.

> should shares have any processing done?
I think no. A user can revoke a granted share, but we assume we handle grants and revokes correctly because it's in the core part of the system.

Reply:
> I could set multiple successfully processed events for the same UD to cleanup_ready, and get the multiple uninstalls back.

That seems like it would be intentionally trying to break the process?  Since there would only ever be one `FAILED` event for any UD there would be no reason to set more than one to `CLEANUP_READY`.

> Say there are "install A", "share A". I set "share A" to clean up ready, run the cleaner, then run the sync. I think this will leave A uninstalled, so the sync does not bring the database into a correct state.

Just had a discussion about this with Steve and there is a piece of the description in the google doc that I failed to capture.  Basically the only time an event should be "uninstalled" is if it was an install event to begin with.  All other event types just get marked with "cleanup_complete" with no action taken.